dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+    dims = sig.split(',')
+    ret = [_parse_dim(d) for d in dims]
+    return tuple(r for r in ret if r is not None)
+
+
+def _parse_args(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^(\([_a-zA-Z0-9,]*\)),?(,?\([_a-zA-Z0-9,]*\))*$", sig)) == 1
+    sigs = re.findall("\([_a-zA-Z0-9,]*\)*", sig)
+    ret = [_parse_arg(s) for s in sigs]
+    return list(r for r in ret if r is not None)
+
+
+def parse_signature(signature):
+    signature = signature.replace(' ', '')
+    assert "->" in signature
+    _in, _out = signature.split("->")
+    _ins = _parse_args(_in)
+    if _ins is None:
+        _ins = []
+    _ins = [i for i in _ins if i is not None]
+    _outs = _parse_args(_out)
+    if _outs and (len(_outs) == 1) and ('),' not in _out):
+        _outs = _outs[0]
+    return _ins, _outs
+
+
+def compile_signature(ins, outs):
+    for e in concat(ins):
+        if len(_valid_name_re.findall(e)) != 1:
+            raise ValueError('`"{0}"` is not a valid name.'.format(e))
+    ins = ",".join(("({0})".format(",".join(dims)) for dims in ins))
+
+    if outs is None:
+        outs = ""
+    elif isinstance(outs, list):
+        for e in concat(outs):
+            if len(_valid_name_re.findall(e)) != 1:
+                raise ValueError('`"{0}"` is not a valid name.'.format(e))
+        _flag = len(outs) == 1
+        outs = ",".join(("({0})".format(",".join(dims)) for dims in outs))
+        if _flag:
+            outs += ","
+    else:
+        for e in outs:
+            if len(_valid_name_re.findall(e)) != 1:
+                raise ValueError('`"{0}"` is not a valid name.'.format(e))
+        outs = "({0})".format(",".join(outs))
+    return "{0}->{1}".format(ins, outs)
+
+
+def apply_gufunc(func, signature, *args, **kwargs):
+    """
+    Apply a generalized ufunc [2]_ to arrays. The function is
+    mapped to the input arguments. The ``signature`` determines
+    if the function consumes or produces core dimensions. The
+    remaining dimensions in given input arrays (``*args``) are
+    considered loop dimensions and are required to broadcast
+    naturally against each other.
+
+    In other terms, this function is like np.vectorize, but for
+    the blocks of dask arrays. If the function itself shall also
+    be vectorized, ``vectorize=True`` can be used for convenience.
+
+    Parameters
+    ----------
+    func : callable
+        Function to call like ``func(*args, **kwargs)`` on input arrays
+        (``*args``) that returns an array or tuple of arrays. If multiple
+        arguments with non-matching dimensions are supplied, this function is
+        expected to vectorize (broadcast) over axes of positional arguments in
+        the style of NumPy universal functions [1]_ (if this is not the case,
+        set ``vectorize=True``).
+    signature: String
+        Specifies what core dimensions are consumed and produced by ``func``.
+        According to the specification of numpy.gufunc signature [2]_
```

Comment:
Quick question, do numpy gufuncs have signatures attached to them?  If so would it make sense to get the signature from the callable?